File: gren/v1/graphs/reducers/markdown-table.js
// Deterministic comparison matrix: rows = records, columns = fields. input: { items: [...], fields?: [...] } args: { fields?: [...] }
export default async function markdownTable(input, args) {
  const items = Array.isArray(input.items) ? input.items : [];
  const fields = Array.isArray(input.fields) ? input.fields : Array.isArray(args.fields) ? args.fields : Object.keys(items[0] ?? {});
  const cell = (v) => (v === undefined || v === null ? "" : typeof v === "object" ? JSON.stringify(v) : String(v)).replace(/\|/g, "\\|").replace(/\r?\n/g, " ");
  const lines = [`| ${fields.join(" | ")} |`, `| ${fields.map(() => "---").join(" | ")} |`];
  for (const it of items) lines.push(`| ${fields.map((f) => cell(it?.[f])).join(" | ")} |`);
  return { markdown: lines.join("\n"), rows: items.length, columns: fields.length, _stats: { in: items.length, out: items.length } };
}

File: gren/kit/graphs/reducers/markdown-table.py
# Deterministic comparison matrix: rows = records, columns = fields. input: { items: [...], fields?: [...] } args: { fields?: [...] }
import json
import re


def _cell(v) -> str:
    if v is None:
        s = ""
    elif isinstance(v, (dict, list)):
        s = json.dumps(v, ensure_ascii=False, default=str)
    else:
        s = str(v)
    return re.sub(r"\r?\n", " ", s.replace("|", "\\|"))


def reduce(input, args, ctx):
    items = input.get("items") if isinstance(input.get("items"), list) else []
    fields = input.get("fields") if isinstance(input.get("fields"), list) else args.get("fields") if isinstance(args.get("fields"), list) else list((items[0] if items and isinstance(items[0], dict) else {}).keys())
    lines = ["| " + " | ".join(str(f) for f in fields) + " |", "| " + " | ".join("---" for _ in fields) + " |"]
    for it in items:
        lines.append("| " + " | ".join(_cell(it.get(f) if isinstance(it, dict) else None) for f in fields) + " |")
    return {"markdown": "\n".join(lines), "rows": len(items), "columns": len(fields), "_stats": {"in": len(items), "out": len(items)}}
